File: src/main/java/es/joaquin/ecommerce/domain/ProductUseCase.java
package es.joaquin.ecommerce.domain;

import java.util.List;
import java.util.Optional;

import es.joaquin.ecommerce.dto.ProductDto;

public interface ProductUseCase {
	
	ProductDto createProduct(ProductDto productDto);

	List<ProductDto> getProducts();

	Optional<ProductDto> getProduct(Long id);

	Boolean deleteProduct(Long id);

}

File: src/domain/productUseCase.js
function init({
  productRepository,
}) {

    async function getProducts() {
      return await productRepository.getProducts();
    }
  
    async function createProduct(productDto) {
      const product = {...productDto};
      return await productRepository.addProduct(product);
    }
  
    async function getProduct(id) {
      return await productRepository.getProduct(id);
    }
  
    async function deleteProduct(id) {
      const productDto = await this.findById(id);
      if (productDto) {
        await productRepository.delete(productDto)
        return productDto;
      }
      return false;
    }

    return {getProducts, createProduct, getProduct, deleteProduct};  
}

module.exports = {init};